## "Pre-lecture" HW

ChatGPT Transcript: https://chatgpt.com/share/46942408-9a13-4eae-81ee-81d600f05ef0

### 1. Pick one of the datasets from the ChatBot session(s) of the **TUT demo** (or from your own ChatBot session if you wish) and use the code produced through the ChatBot interactions to import the data and confirm that the dataset has missing values

In [1]:
import pandas as pd

# URL of the CSV file
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"

# Read the CSV file into a DataFrame
df = pd.read_csv(url)

In [2]:
# Check for missing values in the DataFrame
missing_values = df.isnull().sum()

# Print missing values for each column
print(missing_values)

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64


### 2. Start a new ChatBot session with an initial prompt introducing the dataset you're using and request help to determine how many columns and rows of data a `pandas` DataFrame has, and then

#####    1. use code provided in your ChatBot session to print out the number of rows and columns of the dataset; and,  
#####    2. write your own general definitions of the meaning of "observations" and "variables" based on asking the ChatBot to explain these terms in the context of your dataset

In [3]:
# Get the number of rows and columns
num_rows, num_cols = df.shape

# Print the number of rows and columns
print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_cols}")

Number of rows: 391
Number of columns: 11


In [4]:
df.head()

row_n       id     name  gender    species birthday personality  \
0      2  admiral  Admiral    male       bird     1-27      cranky   
1      3  agent-s  Agent S  female   squirrel      7-2       peppy   
2      4    agnes    Agnes  female        pig     4-21        uchi   
3      6       al       Al    male    gorilla    10-18        lazy   
4      7  alfonso  Alfonso    male  alligator      6-9        lazy   

          song    phrase           full_id  \
0   Steep Hill   aye aye  villager-admiral   
1      DJ K.K.  sidekick  villager-agent-s   
2   K.K. House   snuffle    villager-agnes   
3   Steep Hill   Ayyeeee       villager-al   
4  Forest Life  it'sa me  villager-alfonso   

                                                 url  
0  https://villagerdb.com/images/villagers/thumb/...  
1  https://villagerdb.com/images/villagers/thumb/...  
2  https://villagerdb.com/images/villagers/thumb/...  
3  https://villagerdb.com/images/villagers/thumb/...  
4  https://villagerdb.com/images/villagers/thumb/...

 1. `observations` : the informations abot all villagers.
 2. `variables` : row_n, id, name, gender, species, birthday, personality, song, phrase, full_id, and url

### 3. Ask the ChatBot how you can provide simple summaries of the columns in the dataset and use the suggested code to provide these summaries for your dataset

In [5]:
# Summary of all columns (numeric and non-numeric)
full_summary = df.describe(include='all')
print(full_summary)

             row_n       id     name gender species birthday personality  \
count   391.000000      390      391    391     391      391         391   
unique         NaN      390      391      2      35      361           8   
top            NaN  admiral  Admiral   male     cat     1-27        lazy   
freq           NaN        1        1    204      23        2          60   
mean    239.902813      NaN      NaN    NaN     NaN      NaN         NaN   
std     140.702672      NaN      NaN    NaN     NaN      NaN         NaN   
min       2.000000      NaN      NaN    NaN     NaN      NaN         NaN   
25%     117.500000      NaN      NaN    NaN     NaN      NaN         NaN   
50%     240.000000      NaN      NaN    NaN     NaN      NaN         NaN   
75%     363.500000      NaN      NaN    NaN     NaN      NaN         NaN   
max     483.000000      NaN      NaN    NaN     NaN      NaN         NaN   

                song   phrase           full_id  \
count            380      391       

### 4. If the dataset you're using has (a) non-numeric variables and (b) missing values in numeric variables, explain (perhaps using help from a ChatBot if needed) the discrepancies between size of the dataset given by `df.shape` and what is reported by `df.describe()` with respect to (a) the number of columns it analyzes and (b) the values it reports in the "count" column

(a) df.shape return the total number of rows and columns in df (391 rows, 11 columns), while df.describe() somtimes return less number of rows and columns(Ex: the song column has only 380 entries with actual values (out of 391 rows))

(b)the values it reports in the "count" column is less than the total number of df.

### 5. Use your ChatBot session to help understand the difference between the following and then provide your own paraphrasing summarization of that difference

##### - an "attribute", such as `df.shape` which does not end with `()`
##### - and a "method", such as `df.describe()` which does end with `()` 

1. Attribute : a value or property of an object, and isn't callable.(Doesn't perform any action)

2. Methods : a function, which can perform action to modify the object or return new values.

## "Post-lecture" HW

### 6. The `df.describe()` method provides the 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max' summary statistics for each variable it analyzes. Give the definitions (perhaps using help from the ChatBot if needed) of each of these summary statistics

1. count: Number of valid (non-missing) entries.
2. mean: Average of the data.
3. std: Standard deviation (spread of the data).
4. min: Smallest value in the column.
5. 25%: First quartile (25th percentile).
6. 50%: Median (50th percentile).
7. 75%: Third quartile (75th percentile).
8. max: Largest value in the column.

### 7. Missing data can be considered "across rows" or "down columns".  Consider how `df.dropna()` or `del df['col']` should be applied to most efficiently use the available non-missing data in your dataset and briefly answer the following questions in your own words

##### 1. Provide an example of a "use case" in which using `df.dropna()` might be peferred over using `del df['col']`<br><br>
    
##### 2. Provide an example of "the opposite use case" in which using `del df['col']` might be preferred over using `df.dropna()` <br><br>
    
##### 3. Discuss why applying `del df['col']` before `df.dropna()` when both are used together could be important<br><br>
    
##### 4. Remove all missing data from one of the datasets you're considering using some combination of `del df['col']` and/or `df.dropna()` and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.

1. A dataset of customer information where only a few rows have missing data, it is better to remove only those incomplete rows.
2. If a specific column has a large proportion of missing values, and it is not essential for the analysis
3. If a column has many missing values, it can cause unnecessary data loss when using df.dropna(). Deleting such columns first allows df.dropna() to be applied more efficiently, preserving valuable information in the remaining, mostly complete columns.
4. Only a small portion of rows are missing data. Dropping these rows will remove a small fraction of the data, so the overall integrity of the dataset is maintained.

In [6]:
# Remove rows with any missing data
df_cleaned = df.dropna()

# Check the new shape of the dataset
print("Before cleaning: ", df.shape)
print("After cleaning: ", df_cleaned.shape)

Before cleaning:  (391, 11)
After cleaning:  (379, 11)


### 8. Give brief explanations in your own words for any requested answers to the questions below

1. Use your ChatBot session to understand what `df.groupby("col1")["col2"].describe()` does and then demonstrate and explain this using a different example from the "titanic" data set other than what the ChatBot automatically provide for you

df.groupby("col1"): groups the DataFrame by the unique values in column col1. Each unique value in col1 becomes a group.

["col2"]: This selects column col2 from each group

.describe(): This computes descriptive statistics for column col2 within each group.

In [1]:
# Load the Titanic dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

# Group by 'class' and describe 'age'
result = df.groupby('class')['age'].describe()
print(result)

NameError: name 'pd' is not defined

Google:https://stackoverflow.com/questions/40407090/nameerror-name-pd-is-not-defined

ChatGPT:https://chatgpt.com/share/2cabdbde-07b6-4114-ae37-119ea9960d93

Summary: I think ChatGPT is better, and more efficient. It focuses more on the specific question.

In [2]:
import pandas as pd

# Load the Titanic dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

# Group by 'class' and describe 'age'
result = df.groupby('class')['age'].describe()
print(result)


        count       mean        std   min   25%   50%   75%   max
class                                                            
First   186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
Second  173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
Third   355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0


2. Assuming you've not yet removed missing values in the manner of question "7" above, `df.describe()` would have different values in the `count` value for different data columns depending on the missingness present in the original data.  Why do these capture something fundamentally different from the values in the `count` that result from doing something like `df.groupby("col1")["col2"].describe()`?

df.describe(): Counts non-missing values across the entire DataFrame for each column.

df.groupby("col1")["col2"].describe(): Counts non-missing values within each group defined by col1 for col2.

### 9. Have you reviewed the course [wiki-textbook](https://github.com/pointOfive/stat130chat130/wiki) and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?

yes